src/Graphics/SpriteBatch.cs

```diff
@@ -1067,17 +1068,33 @@ byte effects
 		) {
 			if (numSprites >= vertexInfo.Length)
 			{
-				/* We're out of room, add another batch max
-				 * to the total array size. This is required for
-				 * sprite sorting accuracy; note that we do NOT
-				 * increase the graphics buffer sizes!
-				 * -flibit
-				 */
-				int newMax = vertexInfo.Length + MAX_SPRITES;
-				Array.Resize(ref vertexInfo, newMax);
-				Array.Resize(ref textureInfo, newMax);
-				Array.Resize(ref spriteInfos, newMax);
-				Array.Resize(ref sortedSpriteInfos, newMax);
+				if (vertexInfo.Length >= MAX_BATCHSIZE)
+				{
+					if (sortMode != SpriteSortMode.Immediate)
+					{
+						FlushBatch();
+					}
+				}
+				else
+				{
+					/* We're out of room, add another batch max
+					 * to the total array size. This is required for
+					 * sprite sorting accuracy; note that we do NOT
+					 * increase the graphics buffer sizes!
+					 * -flibit
+					 */
+
+					 /* Fix: reallocation size now increase exponentially, 
+					  * but upper bounded to 1048576. When we submit more than 
+					  * 1048576 sprites, we flush them immediately. This will make
+					  * the total running time O(n).
+					 */ 
+					int newMax = Math.Min(vertexInfo.Length * 2, MAX_BATCHSIZE);
```

Comment:
This, however, is your actual fix, and is worth pushing up to FNA itself, because it's a difference between FNA and XNA impl (XNA does the doubling like you see here, which gives you `log_2(n)` resizes rather than `n`

Reply:
Ok, I'll try to push to FNA first. I'm setting this upper bound because I don't want it to consume too much memory.